language/move-lang/stdlib/modules/libra_account.move

```diff
@@ -106,113 +151,208 @@ module LibraAccount {
-        // Abort if the sender has delegated the privilege to rotate her key elsewhere
-        if (sender_account.delegated_key_rotation_capability) abort 11;
-
+        // The sender has delegated the privilege to rotate her key elsewhere--abort
+        Transaction::assert(!sender_account.delegated_key_rotation_capability, 11);
         // The sender has retained her key rotation privileges--proceed.
-        sender_account.authentication_key = new_authentication_key
+        rotate_authentication_key_for_account(
+            sender_account,
+            new_authentication_key
+        );
     }
 
     // Rotate the authentication key for the account under cap.account_address
     public fun rotate_authentication_key_with_capability(
         cap: &KeyRotationCapability,
         new_authentication_key: bytearray,
     ) acquires T  {
-        borrow_global_mut<T>(cap.account_address).authentication_key = new_authentication_key
+        rotate_authentication_key_for_account(
+            borrow_global_mut<T>(*&cap.account_address),
+            new_authentication_key
+        );
     }
 
     // Return a unique capability granting permission to rotate the sender's authentication key
     public fun extract_sender_key_rotation_capability(): KeyRotationCapability acquires T {
-        let account_address = Transaction::sender();
-        let sender_account = borrow_global_mut<T>(account_address);
+        let sender = Transaction::sender();
+        let sender_account = borrow_global_mut<T>(sender);
         // Abort if we already extracted the unique key rotation capability for this account.
-        if (sender_account.delegated_key_rotation_capability) abort 11;
-
-        // ensure uniqueness of the capability
-        sender_account.delegated_key_rotation_capability = true;
-        KeyRotationCapability { account_address }
+        Transaction::assert(!sender_account.delegated_key_rotation_capability, 11);
+        sender_account.delegated_key_rotation_capability = true; // Ensure uniqueness of the capability
+        KeyRotationCapability { account_address: sender }
     }
 
     // Return the key rotation capability to the account it originally came from
-    public fun restore_key_rotation_capability(cap: Self::KeyRotationCapability) acquires T {
+    public fun restore_key_rotation_capability(cap: KeyRotationCapability) acquires T {
+        // Destroy the capability
         let KeyRotationCapability { account_address } = cap;
+        let account = borrow_global_mut<T>(account_address);
         // Update the flag for `account_address` to indicate that the capability has been restored.
         // The account owner will now be able to call rotate_authentication_key and
         // extract_sender_key_rotation_capability again
-        borrow_global_mut<T>(account_address).delegated_key_rotation_capability = false
+        account.delegated_key_rotation_capability = false;
```

Comment:
I think this is a property that should be encoded as a verifiable property. At the same time, having an assertion there as another statement of this fact shouldn't hurt. However, I'd prefer not to add it to this PR since I want to keep as much semantic similarity to the IR until we switch.